ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -49,21 +73,42 @@ public long getSize() {
 
   @Override
   public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
-    Future<Long> future;
-    if (!channelOpen.get()) {
-      ClosedChannelException closedChannelException = new ClosedChannelException();
-      FutureResult<Long> futureResult = new FutureResult<Long>();
-      futureResult.done(0L, closedChannelException);
-      future = futureResult;
-      if (callback != null) {
-        callback.onCompletion(0L, closedChannelException);
+    ReadIntoCallbackWrapper callbackWrapper = new ReadIntoCallbackWrapper(callback);
+    contentLock.lock();
```

Comment:
don't need a lock if you are not doing `addContent()`. An `AtomicBoolean` and the `compareAndSet` method should do.